File: project01/board.py
#!/usr/bin/env python3

"""
Basic framework for developing 2048 programs in Python

Author: Hung Guei (moporgic)
        Computer Games and Intelligence (CGI) Lab, NCTU, Taiwan
        http://www.aigames.nctu.edu.tw
"""


class board:
    """ simple implementation of Threes puzzle """

    SCORES = [0, 0, 0, 3, 9, 27, 81, 243, 729, 2187, 6561, 19683, 59049, 177147, 531441]
    TILE_VALUES = [0, 1, 2, 3, 6, 12, 24, 48, 96, 192, 384, 768, 1536, 3072, 6144]

    def __init__(self, input_state=None):
        self.state = input_state[:] if input_state is not None else [0] * 16
        return

    def __getitem__(self, pos):
        return self.state[pos]

    def __setitem__(self, pos, tile):
        self.state[pos] = tile
        return

    def place(self, pos, tile):
        """
        place a tile (index value) to the specific position (1-d form index)
        return 0 if the action is valid, or -1 if not
        """
        if pos >= 16 or pos < 0:
            return -1
        if tile < 1 and tile > 3:
            return -1
        self.state[pos] = tile
        return 0

    def slide(self, opcode):
        """
        apply an action to the board
        return the reward of the action, or -1 if the action is illegal
        """
        if opcode == 0:
            return self.slide_up()
        if opcode == 1:
            return self.slide_right()
        if opcode == 2:
            return self.slide_down()
        if opcode == 3:
            return self.slide_left()
        return -1

    def curr_score(self):
        res = 0
        for i in range(16):
            res += self.SCORES[self.state[i]]
        return res

    def num_spaces(self):
        res = 0
        for i in range(16):
            res += 1 if self.state[i] == 0 else 0
        return res

    def slide_left(self):
        new_state, score = [], 0
        for row in [self.state[r:r + 4] for r in range(0, 16, 4)]:
            for i in range(1, 4):
                if row[i] == 0:
                    continue
                if row[i - 1] == 0:
                    row[i - 1] = row[i]
                    row[i] = 0
                elif (row[i - 1] == row[i] and row[i - 1] >= 3 and row[i] >= 3) or \
                     (self.TILE_VALUES[row[i - 1]] + self.TILE_VALUES[row[i]] == 3):

                    row[i - 1] = max(row[i - 1], row[i]) + 1
                    row[i] = 0

            new_state += row

        if self.state == new_state:
            return -1

        self.state = new_state
        return self.curr_score() + self.num_spaces()

    def slide_right(self):
        self.reflect_horizontal()
        score = self.slide_left()
        self.reflect_horizontal()
        return score

    def slide_up(self):
        self.transpose()
        score = self.slide_left()
        self.transpose()
        return score

    def slide_down(self):
        self.transpose()
        score = self.slide_right()
        self.transpose()
        return score

    def reflect_horizontal(self):
        self.state = [self.state[r + i] for r in range(0, 16, 4) for i in reversed(range(4))]
        return

    def reflect_vertical(self):
        self.state = [self.state[c + i] for c in reversed(range(0, 16, 4)) for i in range(4)]
        return

    def transpose(self):
        self.state = [self.state[r + i] for i in range(4) for r in range(0, 16, 4)]
        return

    def rotate(self, rot=1):
        rot = ((rot % 4) + 4) % 4
        if rot == 1:
            self.rotate_right()
            return
        if rot == 2:
            self.reverse()
            return
        if rot == 3:
            self.rotate_left()
            return
        return

    def rotate_right(self):
        """ clockwise rotate the board """
        self.transpose()
        self.reflect_horizontal()
        return

    def rotate_left(self):
        """ counterclockwise rotate the board """
        self.transpose()
        self.reflect_vertical()
        return

    def reverse(self):
        self.reflect_horizontal()
        self.reflect_vertical()
        return

    def __str__(self):
        curr_state = '+' + '-' * 24 + '+\n'
        for row in [self.state[r:r + 4] for r in range(0, 16, 4)]:
            curr_state += ('|' + ''.join('{0:6d}'.format(self.TILE_VALUES[t]) for t in row) + '|\n')
        curr_state += '+' + '-' * 24 + '+'
        return curr_state


if __name__ == '__main__':
    print('2048 Demo: board.py\n')

    state = board()
    state[10] = 10
    print(state)

File: project02/board.cpp
#include <iostream>
#include "board.h"

#define debug(a) std::cout << #a << " = " << a << std::endl
#define print(a) std::cout << a << std::endl

const Board::Reward Board::kTileScore[15] = {0, 0, 0, 3, 9, 27, 81, 243, 729, 2187, 6561, 19683, 59049, 177147, 531441};
const Board::Cell Board::kTileValue[15] = {0, 1, 2, 3, 6, 12, 24, 48, 96, 192, 384, 768, 1536, 3072, 6144};
unsigned int Board::pre_left[66000];
unsigned int Board::pre_score[66000];

/**
 * place a tile (index value) to the specific position (1-d form index)
 * return 0 if the action is valid, or -1 if not
 */
Board::Reward Board::place(unsigned pos, Board::Cell tile_id) {
    if (pos >= 16) return -1;
    if (get_cell(pos) != 0) return -1;
    if (tile_id != 1 && tile_id != 2 && tile_id != 3) return -1;
    set_cell(pos, tile_id);
    board_score += kTileScore[tile_id];
    return 0;
}

/**
 * apply an action to the board
 * return the reward of the action, or -1 if the action is illegal
 */
Board::Reward Board::slide(unsigned opcode) {
    switch (opcode & 0b11u) {
        case 0: return slide_up();
        case 1: return slide_right();
        case 2: return slide_down();
        case 3: return slide_left();
        default: return -1;
    }
}

Board::Reward Board::get_curr_score() const {
    return board_score;
}

Board::Reward Board::slide_left() {

//    debug(*this);

    Board prev = *this;
    board_score = 0;

    for(int i = 0; i < 4; i++) {
        Row row = get_row(i);
        row = ((row & 0xf000ull) >> 12ull) | ((row & 0x0f00ull) >> 4ull) |
                ((row & 0x00f0ull) << 4ull) | ((row & 0x000full) << 12ull);

        Row new_row = pre_left[row];
        board_score += pre_score[new_row];

        new_row = ((new_row & 0xf000ull) >> 12ull) | ((new_row & 0x0f00ull) >> 4ull) |
                  ((new_row & 0x00f0ull) << 4ull) | ((new_row & 0x000full) << 12ull);
        set_row(i, new_row);
    }

//    debug(*this);

    return (*this != prev) ? board_score : -1;
}

Board::Reward Board::slide_right() {
    reflect_horizontal();
    Reward slide_score = slide_left();
    reflect_horizontal();
    return slide_score;
}
Board::Reward Board::slide_up() {
    transpose();
    Reward slide_score = slide_left();
    transpose();
    return slide_score;
}

Board::Reward Board::slide_down() {
    rotate_right();
    Reward slide_score = slide_left();
    rotate_left();
    return slide_score;
}

void Board::transpose() {
    tile = (tile & 0xf0f00f0ff0f00f0full) | ((tile & 0x0000f0f00000f0f0ull) << 12) | ((tile & 0x0f0f00000f0f0000ull) >> 12);
    tile = (tile & 0xff00ff0000ff00ffull) | ((tile & 0x00000000ff00ff00ull) << 24) | ((tile & 0x00ff00ff00000000ull) >> 24);
}

void Board::reflect_horizontal() {

//    debug(*this);
    tile = ((tile & 0x000f000f000f000full) << 12) | ((tile & 0x00f000f000f000f0ull) << 4)
          | ((tile & 0x0f000f000f000f00ull) >> 4) | ((tile & 0xf000f000f000f000ull) >> 12);
//    debug(*this);
}

void Board::reflect_vertical() {

//    debug(*this);
    tile = ((tile & 0x000000000000ffffull) << 48) | ((tile & 0x00000000ffff0000ull) << 16)
          | ((tile & 0x0000ffff00000000ull) >> 16) | ((tile & 0xffff000000000000ull) >> 48);
//    debug(*this);
}

/**
 * rotate the board clockwise by given times
 */
void Board::rotate(int r) {
    switch (((r % 4) + 4) % 4) {
        default:
        case 0: break;
        case 1: rotate_right(); break;
        case 2: reverse(); break;
        case 3: rotate_left(); break;
    }
}

void Board::rotate_right() { transpose(); reflect_horizontal(); } // clockwise
void Board::rotate_left() { transpose(); reflect_vertical(); } // counterclockwise
void Board::reverse() { reflect_horizontal(); reflect_vertical(); }

std::ostream& operator <<(std::ostream& out, const Board& b) {
    out << std::endl;
    out << "+------------------------+" << std::endl;
    for (int i = 0; i < 4; i++) {
        out << "|" << std::dec;
        for (int t = i * 4; t < (i + 1) * 4; t++) out << std::setw(6) << (Board::kTileValue[b.get_cell(t)]);
        out << "|" << std::endl;
    }
    out << "+------------------------+" << std::endl;
    return out;
}

bool Board::can_merge(Board::Cell cell01, Board::Cell cell02) {
    if (cell01 == 1 && cell02 == 2) return true;
    if (cell01 == 2 && cell02 == 1) return true;
    return cell01 > 2 && cell02 > 2 && cell01 == cell02;

}

Board::Cell Board::get_cell(unsigned long long i) const {
    return (tile >> (i << 2ull)) & 0x0full;
}

void Board::set_cell(unsigned long long i, Board::Cell value) {
    tile = (tile & ~(0x0full << (i << 2ull))) | (uint64_t(value & 0x0full) << (i << 2ull));
}

Board::Row Board::get_row(unsigned long long i) const {
    return (tile >> (i << 4ull)) & 0xffffull;
}

void Board::set_row(unsigned long long i, Board::Row value) {
    tile = (tile & ~(0xffffull << (i << 4ull))) | (uint64_t(value & 0xffffull) << (i << 4ull));
}

void Board::precompute_left() {

    for(unsigned int a0 = 0; a0 < 15; a0++) {
        for(unsigned int a1 = 0; a1 < 15; a1++) {
            for(unsigned int a2 = 0; a2 < 15; a2++) {
                for(unsigned int a3 = 0; a3 < 15; a3++) {

                    unsigned int row[4] = {a0, a1, a2, a3};
                    unsigned long long from = row[3] + (row[2] << 4ull) + (row[1] << 8ull) + (row[0] << 12ull);

                    for (int c = 0; c < 3; c++) {
                        if (Board::can_merge(row[c], row[c + 1])) {
                            Cell new_tile = std::max(row[c], row[c + 1]) + 1;
                            row[c] = new_tile;  row[c + 1] = 0;
                        }
                        if (row[c] == 0) {  // can slide
                            row[c] = row[c + 1];
                            row[c + 1] = 0;
                        }
                    }
                    long long score = 0;
                    for (unsigned int cell : row) score += Board::kTileScore[cell];

                    unsigned long long to = row[3] + (row[2] << 4ull) + (row[1] << 8ull) + (row[0] << 12ull);
                    Board::pre_left[from] = to;
                    Board::pre_score[to] = score;
                }
            }
        }
    }
}

Board::Grid Board::get_tile() const {
    return tile;
}







